File: springapp/src/main/java/com/backend/springapp/service/SLAService.java
package com.backend.springapp.service;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.backend.springapp.exception.DuplicateResourceException;
import com.backend.springapp.exception.ResourceNotFoundException;
import com.backend.springapp.model.Category;
import com.backend.springapp.model.Department;
import com.backend.springapp.model.SLA;
import com.backend.springapp.repository.CategoryRepository;
import com.backend.springapp.repository.DepartmentRepository;
import com.backend.springapp.repository.SLARepository;

@Service
@Transactional
public class SLAService {

    @Autowired
    private SLARepository slaRepository;

    @Autowired
    private CategoryRepository categoryRepository;

    @Autowired
    private DepartmentRepository departmentRepository;

    /**
     * Create SLA config for a category (Admin only)
     * Links Category → Department with SLA days and base priority
     */
    public SLA createSLAConfig(Long categoryId, Long departmentId, SLA slaConfig) {
        Category category = categoryRepository.findById(categoryId)
            .orElseThrow(() -> new ResourceNotFoundException("Category not found with id: " + categoryId));

        // Check if SLA already exists for this category
        if (slaRepository.findByCategoryId(categoryId).isPresent()) {
            throw new DuplicateResourceException("SLA config already exists for category: " + category.getName());
        }

        Department department = departmentRepository.findById(departmentId)
            .orElseThrow(() -> new ResourceNotFoundException("Department not found with id: " + departmentId));

        slaConfig.setCategory(category);
        slaConfig.setDepartment(department);

        return slaRepository.save(slaConfig);
    }

    @Transactional(readOnly = true)
    public SLA getSLAById(Long id) {
        return slaRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("SLA config not found with id: " + id));
    }

    @Transactional(readOnly = true)
    public SLA getSLAByCategoryId(Long categoryId) {
        return slaRepository.findByCategoryId(categoryId)
            .orElseThrow(() -> new ResourceNotFoundException("SLA config not found for category id: " + categoryId));
    }

    @Transactional(readOnly = true)
    public SLA getSLAByCategoryName(String categoryName) {
        Category category = categoryRepository.findByName(categoryName)
            .orElseThrow(() -> new ResourceNotFoundException("Category not found with name: " + categoryName));
        return slaRepository.findByCategory(category)
            .orElseThrow(() -> new ResourceNotFoundException("SLA config not found for category: " + categoryName));
    }

    @Transactional(readOnly = true)
    public List<SLA> getAllSLAConfigs() {
        return slaRepository.findAll();
    }

    /**
     * Update SLA config (Admin only)
     * Can change: slaDays, basePriority, department
     */
    public SLA updateSLAConfig(Long id, SLA updatedSLA) {
        SLA existing = slaRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("SLA config not found with id: " + id));

        if (updatedSLA.getSlaDays() != null) {
            existing.setSlaDays(updatedSLA.getSlaDays());
        }
        if (updatedSLA.getBasePriority() != null) {
            existing.setBasePriority(updatedSLA.getBasePriority());
        }

        return slaRepository.save(existing);
    }

    /**
     * Update SLA config's department (Admin only)
     */
    public SLA updateSLADepartment(Long slaId, Long departmentId) {
        SLA existing = slaRepository.findById(slaId)
            .orElseThrow(() -> new ResourceNotFoundException("SLA config not found with id: " + slaId));

        Department department = departmentRepository.findById(departmentId)
            .orElseThrow(() -> new ResourceNotFoundException("Department not found with id: " + departmentId));

        existing.setDepartment(department);
        return slaRepository.save(existing);
    }

    /**
     * Delete SLA config (Admin only)
     */
    public void deleteSLAConfig(Long id) {
        if (!slaRepository.existsById(id)) {
            throw new ResourceNotFoundException("SLA config not found with id: " + id);
        }
        slaRepository.deleteById(id);
    }
}

File: reactapp/src/services/api/slaService.js
/**
 * SLA Service
 * 
 * All SLA-related API calls
 * Uses ONLY endpoints defined in ENDPOINTS.md
 */

import apiClient from './apiClient';

const slaService = {
  /**
   * Get all SLA configs
   * GET /api/sla
   */
  getAll: () => {
    return apiClient.get('/sla');
  },

  /**
   * Create a new SLA
   * POST /api/sla?categoryId={categoryId}&departmentId={departmentId}
   * @param {Object} slaData - { slaDays, basePriority }
   * @param {string|number} categoryId 
   * @param {string|number} departmentId 
   */
  create: (slaData, categoryId, departmentId) => {
    return apiClient.post(`/sla?categoryId=${categoryId}&departmentId=${departmentId}`, slaData);
  },

  /**
   * Get SLA by ID
   * GET /api/sla/{id}
   * @param {string|number} id 
   */
  getById: (id) => {
    return apiClient.get(`/sla/${id}`);
  },

  /**
   * Get SLA by category ID
   * GET /api/sla/category/{categoryId}
   * @param {string|number} categoryId 
   */
  getByCategoryId: (categoryId) => {
    return apiClient.get(`/sla/category/${categoryId}`);
  },

  /**
   * Get SLA by category name
   * GET /api/sla/category/name/{categoryName}
   * @param {string} categoryName 
   */
  getByCategoryName: (categoryName) => {
    return apiClient.get(`/sla/category/name/${encodeURIComponent(categoryName)}`);
  },

  /**
   * Update SLA config
   * PUT /api/sla/{id}
   * @param {string|number} id
   * @param {Object} slaData - { slaDays, basePriority }
   */
  update: (id, slaData) => {
    return apiClient.put(`/sla/${id}`, slaData);
  },

  /**
   * Update SLA department
   * PUT /api/sla/{id}/department/{departmentId}
   * @param {string|number} slaId 
   * @param {string|number} departmentId 
   */
  updateDepartment: (slaId, departmentId) => {
    return apiClient.put(`/sla/${slaId}/department/${departmentId}`);
  },

  /**
   * Delete SLA config
   * DELETE /api/sla/{id}
   * @param {string|number} id
   */
  delete: (id) => {
    return apiClient.delete(`/sla/${id}`);
  },
};

export default slaService;
